api/classes/class-lightweight-api.php

```diff
@@ -227,161 +226,580 @@ public function delete_transient( $name ) {
+	 * Get all reader data from the DB.
+	 *
+	 * @param string $client_id Client ID of the reader.
+	 *
+	 * @return array Array of reader data for the given client ID.
+	 */
+	public function get_reader_events_from_db( $client_id ) {
 		global $wpdb;
-		$events_table_name       = Segmentation::get_events_table_name();
-		$client_post_read_events = $wpdb->get_results( // phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
-			$wpdb->prepare( "SELECT * FROM $events_table_name WHERE client_id = %s AND type = 'post_read'", $client_id ) // phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+		$reader_events_table_name = Segmentation::get_reader_events_table_name();
+		$events                   = $wpdb->get_results( // phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
+			$wpdb->prepare(
+				"SELECT client_id, date_created, type, context, value from $reader_events_table_name WHERE client_id = %s ORDER BY date_created DESC", // phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+				$client_id
+			)
 		);
 
-		// If there is no relevant client data in the events table, do not save any data.
-		if ( 0 === count( $client_post_read_events ) ) {
-			return $this->client_data_blueprint;
+		$this->debug['read_query_count'] += 1;
+
+		if ( ! empty( $events ) ) {
+			$events = array_map(
+				function( $event ) {
+					$event = (array) $event;
+
+					if ( ! empty( $event['value'] ) && ! is_array( $event['value'] ) ) {
+						$event['value'] = json_decode( $event['value'], true );
+					}
+
+					return $event;
+				},
+				$events
+			);
 		}
 
-		$this->save_client_data(
-			$client_id,
-			[
-				'posts_read' => array_map(
-					function ( $item ) {
-						return [
-							'post_id'      => $item->post_id,
-							'category_ids' => $item->category_ids,
-							'created_at'   => $item->created_at,
-						];
-					},
-					$client_post_read_events
-				),
-			]
+		return $events;
+	}
+
+	/**
+	 * Retrieve data for a specific reader from the cache or DB.
+	 *
+	 * @param string            $client_id Client ID of the reader.
+	 * @param string|array|null $type Type or types of data to retrieve.
+	 *                                If not given, only return temporary data types.
+	 * @param string|array|null $context Context or contexts of data to retrieve.
+	 *
+	 * @return array Array of reader data, optionally filtered by $type and $context.
+	 */
+	public function get_reader_events( $client_id, $type = null, $context = null ) {
+		$events = [];
+
+		// Check the cache first.
+		if ( ! $this->ignore_cache ) {
+			$events = wp_cache_get( 'reader_events', $client_id );
+			if ( ! empty( $events ) ) {
+				$this->debug['reader_events'] = $events;
+				return $this->filter_events_by_type( $events, $type, $context );
+			}
+		}
+
+		// If no data in the cache, get it from the DB.
+		$events = $this->get_reader_events_from_db( $client_id );
```

Comment:
Storing the reader events with proper structure and indexes allows us to have a more powerful API to perform optimized queries. Shouldn't we start using those queries instead of always fetching all events and processing/filtering on the PHP?

Reply:
In that case, it is more costly to filter the data on PHP than to use indexed and optimized queries to fetch the exact data that is going to be used.

The number of DB hits should be the same, with the difference that we'd be skipping a PHP processing/filtering step, making it generally lighter. Fewer items to retrieve and fewer data to process.

We can optimize that on another PR, but I'd still prioritize it for performance sake.

Reply:
I [created an issue](https://github.com/Automattic/newspack-popups/issues/907) to tackle this in a separate PR because I don't want to refactor this data handling again in this PR!